test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java

```diff
@@ -0,0 +1,221 @@
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 + inv2 == i`
+            if (i - inv1 == inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.ADD_L, "1"})
+    public void equalsVariantSubInvariantLong(long inv1, long inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 + inv2 == i`
+            if (i - inv1 == inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.SUB_I, "1"})
+    public void notEqualsAddInt(int inv1, int inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 - inv2 != i`
+            if (inv1 + i != inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.SUB_L, "1"})
+    public void notEqualsAdtLong(long inv1, long inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 - inv2 != i`
+            if (inv1 + i != inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.SUB_I, "1"})
+    public void notEqualsInvariantSubVariantInt(int inv1, int inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv2 - inv1 != i`
+            if (inv1 - i != inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.SUB_L, "1"})
+    public void notEqualsInvariantSubVariantLong(long inv1, long inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv2 - inv1 != i`
+            if (inv1 - i != inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.ADD_I, "2"})
+    public void notEqualsVariantSubInvariantInt(int inv1, int inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 + inv2 != i`
+            if (i - inv1 != inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.ADD_L, "1"})
+    public void notEqualsVariantSubInvariantLong(long inv1, long inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 + inv2 != i`
+            if (i - inv1 != inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(failOn = {IRNode.SUB_I})
+    public void leDontReassociate(int inv1, int inv2) {
```

Comment:
I added DontReassociate tests for `le`, `gt`, and `ge`. For `lt`, C2 generates a second `SUB_I` as part of other transformations.

IR matching for ADD/SUB is pretty hard in general. They commonly are created as part of other transformations. Any suggestions on how I can test this better is appreciated.

Reply:
You could always use a simple regex with the linenumber I guess. But that is a bit nasty too.

Reply:
Yeah, it would work for this patch. But people working on future unrelated changes may have to change the line number. Seems more pain than its worth.

Reply:
Another suboptimal idea: you wrap the add / sub in a method, and then ensure that this method is inlined. It might still keep the annotation of being part of that inner method, and you could use regex to check for it.

Or maybe we could also have some sort of relative line offset mechanism in the IR framework, that allows you to specify that you want something that is let's say 7 lines down from the IR rule.

Reply:
Some sort of pattern matcher could work. It would be able nice to match something like `a ADD_I b CMP_LT c`. In java this could look something like 

```
@IR(counts = {IRNode.CMP_LT[IRNode.ANY, IRNode.SUB_I, IRNode.ANY], "1"}
```

The arguments in the `[]` are the inputs. `IRNode.ANY` matches any node. (The zero'th node is ANY because its the region node).

Anyway, I think a `lt` test is not super-required for the coverage for this PR. The current machinery does not provide a convenient way to test it. I'd prefer to avoid something hacky. I think this work can be done separately.

Reply:
I agree with you there, don't do anything hacky here.
But yes, I've also been wondering what kind of improvements to the IR framework would help us to do these sorts of graph-matching verifications.

Reply:
Created https://bugs.openjdk.org/browse/JDK-8324226